### PR TITLE
Shims should use intptr_t for file descriptors.

### LINKED ISSUE #2904: Shims should use intptr_t for file descriptors
Once we get all of the IO stuff shimmed, it would be better to use intptr_t instead of int32_t for file descriptors as then we can marshal SafeHandles to descriptors without the current workarounds.

### PR DESCRIPTION
**Pull request by eerhardt**
Changing all file descriptors used in System.Native to use intptr_t instead of int32_t.  This way we can marshal SafeHandles without having to add DangerousAddRef/Release code.

Fix #2904.

@nguerrera @stephentoub @sokket 

@pgavlin - can you take a look at the Sockets changes?

### CODE REVIEW

File: src/Common/src/Interop/Unix/System.Native/Interop.FileDescriptors.cs [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
     internal static partial class Sys
     {
         internal static class FileDescriptors
         {
-            internal const int STDIN_FILENO = 0;
-            internal const int STDOUT_FILENO = 1;
-            internal const int STDERR_FILENO = 2;
+            internal static readonly SafeFileHandle STDIN_FILENO = new SafeFileHandle((IntPtr)0, false);
+            internal static readonly SafeFileHandle STDOUT_FILENO = new SafeFileHandle((IntPtr)1, false);
+            internal static readonly SafeFileHandle STDERR_FILENO = new SafeFileHandle((IntPtr)2, false);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.FileDescriptors.cs]: Nit: could you name the Boolean arguments, e.g. ```ownsHandle: false```?  That helps me as a reader to understand very clearly what it means.

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs [hunk 2]
```diff
@@ -35,7 +40,7 @@ internal static SafeFileHandle Open(string path, Interop.Sys.OpenFlags flags, in
             bool enoentDueToDirectory = (flags & Interop.Sys.OpenFlags.O_CREAT) != 0;
 
             // Open the file. 
-            int fd;
+            IntPtr fd;
```
Comment (stephentoub) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: Should Open be defined to return a SafeFileHandle?  Then we could have an Interop.CheckIo that takes a SafeHandle instead of an IntPtr, and we could avoid doing the whole dance of instantiating the SafeFileHandle and then calling SetHandle on it subsequently.

File: src/Common/src/Interop/Unix/System.Native/Interop.Read.cs [hunk 3]
```diff
@@ -18,6 +19,6 @@ internal static partial class Sys
         /// Note - on fail. the position of the stream may change depending on the platform; consult man 2 read for more info
         /// </returns>
         [DllImport(Libraries.SystemNative, SetLastError = true)]
-        internal static unsafe extern int Read(int fd, byte* buffer, int count);
+        internal static unsafe extern int Read(SafeFileHandle fd, byte* buffer, int count);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.Read.cs]: I think it's ok to have the two different Read declarations, one for SafeFileHandle and one for SafePipeHandle, but you might also consider just having one declaration that takes the base SafeHandle.

File: src/Common/src/Interop/Unix/System.Native/Interop.Stat.Pipe.cs [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+using System;
+using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
+
+internal static partial class Interop
+{
+    internal static partial class Sys
+    {
+        [DllImport(Libraries.SystemNative, SetLastError = true)]
+        internal static extern int FStat(SafePipeHandle fd, out FileStatus output);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.Stat.Pipe.cs]: Ditto on considering using SafeHandle to avoid needing two declarations.

File: src/Common/src/Interop/Unix/System.Native/Interop.Dup.cs [hunk 5]
```diff
@@ -1,13 +1,15 @@
...
 using System.Runtime.InteropServices;
+using Microsoft.Win32.SafeHandles;
 
 internal static partial class Interop
 {
     internal static partial class Sys
     {
         [DllImport(Libraries.SystemNative, SetLastError = true)]
-        internal static extern int Dup(int oldfd);
+        internal static extern IntPtr Dup(SafeFileHandle oldfd);
```
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.Dup.cs]: Should this return a SafeFileHandle instead of an IntPtr?  The SafeFileHandle.Open would then change to use the CheckIo function I suggested that takes a SafeHandle.

File: src/Common/src/Interop/Unix/System.Native/Interop.Read.cs (hunk 3 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Native/Interop.Read.cs]: My thinking/reasoning was type-safety/discoverability.  I know this is internal-only code, but it felt odd that Read could take *any* SafeHandle, i.e. not a File or Pipe handle.

I can change it, if you'd like, but that is where my head was at.

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs (hunk 2 above)
Comment (eerhardt) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: The part that hung me up was the `while (CheckIO(...))` loop.  If the open call gets interrupted a lot, won't we be creating and throwing away a lot of finalizable SafeHandle objects?  Is that a concern?

Comment (stephentoub) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: Hmm.  It shouldn't happen very often (basically when the process receives a signal during the invocation of one of these operations), but I agree if it was more prevalent we'd want to avoid those costs.  Why don't we stick with what you have for now, but then separately do some testing, instrumenting our checks for EINTR to track how often it happens, and then running some workloads to see how often it happens.  I like the cleaniness that your PR here introduces, and it'd be nice to take it as far as possible.

File: src/Common/src/Interop/Unix/System.Native/Interop.Read.cs (hunk 3 above)
Comment (stephentoub) [inline src/Common/src/Interop/Unix/System.Native/Interop.Read.cs]: It's up to you.  If you think the type safety is worthwhile, I'm fine with it.

File: src/Native/System.Native/pal_io.cpp [hunk 6]
```diff
@@ -127,6 +127,12 @@ static_assert(PAL_IN_EXCL_UNLINK == IN_EXCL_UNLINK, "");
 static_assert(PAL_IN_ISDIR == IN_ISDIR, "");
 #endif
 
+inline static int ToFileDescriptor(intptr_t fd)
+{
+    assert(INT32_MIN <= fd && fd <= INT32_MAX);
```
Comment (stephentoub) [inline src/Native/System.Native/pal_io.cpp]: A file descriptor should never be negative.

File: src/Native/System.Native/pal_console.cpp [hunk 7]
```diff
@@ -35,9 +35,10 @@ extern "C" int32_t GetWindowSize(WinSize* windowSize)
 #endif
 }
 
-extern "C" int32_t IsATty(int fd)
+extern "C" int32_t IsATty(intptr_t fd)
 {
-    return isatty(fd);
+    assert(INT32_MIN <= fd && fd <= INT32_MAX);
```
Comment (stephentoub) [inline src/Native/System.Native/pal_console.cpp]: You have a ToFileDescriptor defined elsewhere.  Can this just use that?

File: src/System.Console/src/System/TermInfo.cs [hunk 8]
```diff
@@ -208,20 +209,20 @@ private static Database ReadDatabase(string term)
             /// <param name="filePath">The path to the file to open.</param>
             /// <param name="fd">If successful, the opened file descriptor; otherwise, -1.</param>
             /// <returns>true if the file was successfully opened; otherwise, false.</returns>
-            private static bool TryOpen(string filePath, out int fd)
+            private static bool TryOpen(string filePath, out SafeFileHandle fd)
             {
-                int tmpFd;
-                while ((tmpFd = Interop.Sys.Open(filePath, Interop.Sys.OpenFlags.O_RDONLY, 0)) < 0)
+                IntPtr tmpFd;
+                while ((int)(tmpFd = Interop.Sys.Open(filePath, Interop.Sys.OpenFlags.O_RDONLY, 0)) < 0)
```
Comment (stephentoub) [inline src/System.Console/src/System/TermInfo.cs]: Nit: this shouldn't make any difference, but just in the name of robustness, should this cast be to long rather than int?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 9]
```diff
@@ -87,14 +88,14 @@ private void EnsureWorkerIsRunning()
...
                     // This is created here rather than in Process so that the pipe is available immediately
                     // for subsequent queue calls to use.
-                    Debug.Assert(_wakeupRequestedPipeFd == 0, "Read pipe should have been cleared");
-                    Debug.Assert(_requestWakeupPipeFd == 0, "Write pipe should have been cleared");
+                    Debug.Assert(_wakeupRequestedPipeFd == null, "Read pipe should have been cleared");
+                    Debug.Assert(_requestWakeupPipeFd == null, "Write pipe should have been cleared");
                     unsafe
                     {
                         int* fds = stackalloc int[2];
                         while (Interop.CheckIo(Interop.Sys.Pipe(fds))) ;
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Should our shim over pipe be changed to take an ```intptr_t*``` instead of an ```int*```?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs [hunk 10]
```diff
@@ -293,7 +294,7 @@ private void WorkerLoop()
                         // Wait for more things to do.
                         bool isWakeupRequestedPipeActive;
                         bool isTimeout;
-                        ThrowIfCURLMError(Interop.Http.MultiWait(multiHandle, _wakeupRequestedPipeFd, out isWakeupRequestedPipeActive, out isTimeout));
+                        ThrowIfCURLMError(Interop.Http.MultiWait(multiHandle, (int)_wakeupRequestedPipeFd.DangerousGetHandle(), out isWakeupRequestedPipeActive, out isTimeout));
```
Comment (stephentoub) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: Seems like MultiWait should be changed to have the file descriptor argument be an ```intptr_t```?

File: src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs [hunk 11]
```diff
@@ -20,9 +21,10 @@ internal enum PollEvents : short
 
         internal struct PollEvent
         {
-            internal int FileDescriptor;         // The file descriptor to poll
+            internal IntPtr FileDescriptor;      // The file descriptor to poll
```
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs]: @nguerrera - do you think this one is necessary?  We don't get any of the benefits here, since we can't marshal using a SafeHandle.  And this change causes the size of PollEvent to increase and to require padding.

The only advantage we get is that it is consistent with the rest of the file descriptors.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 9 above)
Comment (eerhardt) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: I skipped this one at first because I didn't think it would add value, and would only add complexity to the shim code.

But thinking more about it, if we are going to say all file descriptors are marshaled as intptr_t, then consistency is the value added.  I can change it.

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 10 above)
Comment (eerhardt) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: I was trying to cut the changes off at pal_io in System.Native, as this change got bigger than I expected. But looking at it now, this would be a pretty easy change to make.

The other place that is left undone is the sockets code - it is still using int32_t for the socket file descriptor.  I can fix that in a subsequent PR if people feel it would be beneficial.   @pgavlin ???

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs (hunk 2 above)
Comment (eerhardt) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: Another option we have here is to move the EINTR checks into the shim (obviously with a different change, not this one).  And then managed code shouldn't need to worry about them at all.  Then I think the managed code could be really clean: Open returns a SafeFileHandle and managed wouldn't even need the while loop.  Just a check to throw on an Invalid handle.

File: src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs (hunk 11 above)
Comment (pgavlin) [inline src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs]: I personally do not think that this is necessary. As you've pointed out, its only benefit is consistency and it does come at a not insignificant cost.

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs (hunk 2 above)
Comment (stephentoub) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: I'd be ok with that.

File: src/Native/System.Native/pal_console.cpp (hunk 7 above)
Comment (pgavlin) [inline src/Native/System.Native/pal_console.cpp]: IMHO these shouldn't be asserts in any case. If we're extending the potential range of inputs, we should be validating that input, setting `errno`, and returning something that indicates failure when necessary.

Comment (pgavlin) [inline src/Native/System.Native/pal_console.cpp]: (note that we could probably rely on the platform to do this for us by collapsing all out-of-range file descriptors to `-1`)

**Comment by pgavlin** [comment]
As usual, I'm a bit dismayed to see us relying more on magic marshalling. I continue to feel that it obscures important operations and makes failure points (and behavior) less clear, esp. in the case of SafeHandles (where a good understanding of the marshalling behavior requires reading at least one or two MSDN articles).

Aside from my fundamental unease with this change, the sockets changes do look OK.

**Comment by stephentoub** [comment]
@pgavlin, you mean you object to using SafeHandles in P/Invoke signatures at all?

**Comment by pgavlin** [comment]
Personally, yes. I understand the justification--it is nice to not need to write the ref counting manually--but especially after my experience with the patterns that using `SafeHandle` in P/Invoke signatures engenders in Sockets (e.g. overallocation, overuse of `SafeHandle.IsInvalid`, etc.) I'm not convinced that the benefits outweigh the drawbacks. Perhaps the benefits are clearer when `SafeHandle`s are confined to parameters (rather than return values) which is largely the case here.

File: src/Native/System.Native/pal_console.cpp (hunk 7 above)
Comment (eerhardt) [inline src/Native/System.Native/pal_console.cpp]: My understanding was that these asserts are to help us catch bugs in our code, and aren't really to validate a "public" contract.

We've treated the shim APIs as internal in other places:  https://github.com/dotnet/corefx/pull/4267#discussion_r43602869

Comment (sokket) [inline src/Native/System.Native/pal_console.cpp]: Also - the maximum FD isn't necessarily INT32_MAX (I believe). On OS X the max is OPEN_MAX in sys/syslimits.h. We should try to use these constants in our asserts if Ubuntu has a similar constant

File: src/Common/src/Interop/Unix/System.Native/Interop.Dup.cs (hunk 5 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Native/Interop.Dup.cs]: This returns an IntPtr for the same reason Open returns an IntPtr - because of the looping EINTR checks.  This should be removed when we move the EINTR checks into the native shim.

File: src/Common/src/Interop/Unix/System.Native/Interop.FileDescriptors.cs (hunk 1 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Native/Interop.FileDescriptors.cs]: Done.  I created a static method to create these SafeFileHandles and added the named parameter.

File: src/Native/System.Native/pal_console.cpp (hunk 7 above)
Comment (eerhardt) [inline src/Native/System.Native/pal_console.cpp]: I did a quick look, but I don't see a good define for this on Linux.  I changed it to use OPEN_MAX when it is defined.

File: src/Native/System.Native/pal_io.cpp (hunk 6 above)
Comment (eerhardt) [inline src/Native/System.Native/pal_io.cpp]: Fixed.

File: src/System.Console/src/System/TermInfo.cs (hunk 8 above)
Comment (eerhardt) [inline src/System.Console/src/System/TermInfo.cs]: Done.

File: src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs (hunk 11 above)
Comment (eerhardt) [inline src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs]: I'm going to revert this change, and along with that not change `Sys.Pipe` to take in IntPtr[].  Both of these changes come at a cost, with minimal benefits.
@stephentoub @nguerrera - Are you guys OK with this?

File: src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs (hunk 9 above)
Comment (eerhardt) [inline src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs]: I decided not to make this change.  See https://github.com/dotnet/corefx/pull/4791/files#r46606957 for the discussion.

**Comment by eerhardt** [comment]
I've responded to all feedback and pushed a new commit.  PTAL.

**Comment by sokket** [comment]
LGTM

File: src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs (hunk 11 above)
Comment (nguerrera) [inline src/Common/src/Interop/Unix/System.Native/Interop.Poll.cs]: Agree that if we can't use SafeHandle, there's no point to the change.

**Comment by nguerrera** [comment]
LGTM.

File: src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs (hunk 2 above)
Comment (nguerrera) [inline src/Common/src/Microsoft/Win32/SafeHandles/SafeFileHandle.Unix.cs]: I like the idea of pushing EINTR handling down to the shims.

**Comment by stephentoub** [comment]
There are a ton of errors in the CI runs due to MissingMethodExceptions.  I wonder if this has something to do with breaking changes to the shims?

**Comment by eerhardt** [comment]
The tests fail with:
`System.MissingMethodException : .ctor`

from both `Interop.Sys.INotifyInit()` and `at Interop.Sys.ShmOpen` for 2 reasons:

1. The SafeFileHandle class didn't have a default constructor before.  So if you try P/Invoking a method that returns a SafeFileHandle you get a MissingMethodException because the runtime doesn't know how to create a SafeFileHandle.
 1. I've fixed this in my "Respond to PR feedback" commit.
2. The reason they are still failing is because of the way we run tests in CI (and during ./build.sh on a dev box).  When we run XXX.Test.dll, we are actually running the tests against the locally built XXX.dll.  However, when XXX.dll depends on YYY.dll, we don't use the locally built YYY.dll when running XXX.Test.dll tests.  Instead, we grab a previously built YYY.dll from nuget.
 1. So in this case, System.IO.FileSystem has the SafeFileHandle class in it.  Both System.IO.FileSystem.Watcher and System.IO.Pipes depend on System.IO.FileSystem.  I added the new constructor to SafeFileHandle in System.IO.FileSystem, but when running the Watcher and Pipes tests, CI is using the System.IO.FileSystem that was already built from sometime in the past.  It is not using the most recent code.  
 2. This can also cause false positives in tests if new code in System.IO.FileSystem actually breaks Watcher or Pipes tests.  Let's say someone would go and remove the default constructor.  All CI tests would pass.  But in actuality, Watcher and Pipes tests would now be broken when they picked up the latest System.IO.FileSystem assembly.

**Comment by stephentoub** [comment]
Yeah, I had a similar problem here: https://github.com/dotnet/corefx/pull/4816 (see the second commit).  You might do something similar as a temporary workaround, but we need to come up with something better for the long run.

**Comment by eerhardt** [comment]
I've added a similar workaround.

**Comment by eerhardt** [comment]
@stephentoub - I needed to make a fix for the file descriptor assert since it was failing in MemoryMappedFiles.  Can you PTAL?

**Comment by eerhardt** [comment]
@stephentoub - I added the new function ToFileDescriptorUnchecked and call it from MMap.  Thanks for the review.
